[Merged by Bors] - Make grpc servers' start methods be synchronous

piersy opened the pull request:
## Motivation
Closes #4861 
## Changes

Updates grpc server startup code to execute synchronously, returning only after the server has started listening.

## Test Plan
Existing tests should pass

## DevOps Notes
<!-- Please uncheck these items as applicable to make DevOps aware of changes that may affect releases -->
- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

piersy committed:
Make grpc servers' start methods be synchronous

fasmat commented on api/grpcserver/grpcserver_test.go:
```diff
@@ -491,20 +491,14 @@ func launchServer(tb testing.TB, cfg Config, services ...ServiceAPI) func() {
+	err = jsonService.StartService(context.Background(), services...)
+	require.NoError(tb, err)
 
 	timer := time.NewTimer(3 * time.Second)
 	defer timer.Stop()
```
I think this is not needed any more 🙂

fasmat commented on cmd/bootstrapper/generator_test.go:
```diff
@@ -139,20 +139,14 @@ func launchServer(tb testing.TB, cdb *datastore.CachedDB) func() {
+	err = jsonService.StartService(context.Background(), s)
+	require.NoError(tb, err)
 
 	timer := time.NewTimer(3 * time.Second)
 	defer timer.Stop()
```
Same here, I don't think this is needed any more.

poszu commented on api/grpcserver/grpc.go:
```diff
@@ -44,35 +46,27 @@ func (s *Server) Start() <-chan struct{} {
 // Close stops the server.
 func (s *Server) Close() error {
 	s.logger.Info("stopping the grpc server")
 	s.GrpcServer.Stop()
```
Consider a graceful stop to allow pending GRPC calls to finish:
```suggestion
	s.GrpcServer.GracefulStop()
```

piersy replied:
I don't think just calling GracefulStop would work well with our streaming APIs, I think if you had a stream open the server would never shutdown. The approach to handling this, according to this [ticket](https://github.com/grpc/grpc-go/issues/2888), is to first call GracefulStop, wait some time and then call Stop. I can add a ticket for it, it seems to be out of scope for this PR.

poszu replied:
Thanks, didn't know that! Do you think it's worth creating an issue to implement the proposed graceful shutdown later?

piersy replied:
Sure will do!

piersy replied:
https://github.com/spacemeshos/go-spacemesh/issues/4867

piersy committed:
Remove useless timer

piersy commented:
bors merge

piersy commented:
bors cancel

piersy committed:
Merge branch 'develop' into piersy/grpcserver-synchronous-start

piersy commented:
bors merge

piersy committed:
Fix multiple problems

Remove erroneous blocking call to serve in grpc http server.
Return error returned by grpc server Start methods.
Fix bad peer test config.

piersy committed:
Fix failing tests

piersy committed:
Merge branch 'develop' into piersy/grpcserver-synchronous-start

piersy commented:
bors merge
